### ISSUE TITLE
Is the result reasonable?

### ISSUE THREAD

**Issue by JohnZhan2023** [SandAI-org/MAGI-1#37]
Thank you for your wonderful work! I downloaded the distilled and fp8_quant version and use the script below, and I get the result like this (using h200):

https://github.com/user-attachments/assets/e20073d7-f1e0-41bb-aa4e-abfb90bd0f4a

I think the video result is a little poor and I wonder whether it's correct.
I change the frame_num in run_ and my config is:
```sh
{
    "model_config": {
        "model_name": "videodit_ardf",
        "num_layers": 48,
        "hidden_size": 6144,
        "ffn_hidden_size": 16384,
        "num_attention_heads": 48,
        "num_query_groups": 8,
        "kv_channels": 128,
        "layernorm_epsilon": 1e-06,
        "apply_layernorm_1p": true,
        "x_rescale_factor": 0.1,
        "half_channel_vae": true,
        "params_dtype": "torch.bfloat16",
        "patch_size": 2,
        "t_patch_size": 1,
        "in_channels": 32,
        "out_channels": 32,
        "cond_hidden_ratio": 0.25,
        "caption_channels": 4096,
        "caption_max_length": 800,
        "xattn_cond_hidden_ratio": 1.0,
        "cond_gating_ratio": 1.0,
        "gated_linear_unit": true
    },
    "runtime_config": {
        "cfg_number": 1,
        "cfg_t_range": [
            0.0,
            0.0217,
            0.1,
            0.3,
            0.999
        ],
        "prev_chunk_scales": [
            1.5,
            1.5,
            1.5,
            1.0,
            1.0
        ],
        "text_scales": [
            7.5,
            7.5,
            7.5,
            0.0,
            0.0
        ],
        "noise2clean_kvrange": [
            5,
            4,
            3,
            2
        ],
        "clean_chunk_kvrange": 1,
        "clean_t": 0.9999,
        "seed": 1234,
        "num_frames": 23,
        "video_size_h": 480,
        "video_size_w": 640,
        "num_steps": 4,
        "window_size": 4,
        "fps": 24,
        "chunk_width": 6,
        "load": "./ckpt/magi/24B_distill_quant",
        "t5_pretrained": "./ckpt/t5",
        "vae_pretrained": "./ckpt/vae",
        "scale_factor": 0.18215,
        "temporal_downsample_factor": 4
    },
    "engine_config": {
        "distributed_backend": "nccl",
        "distributed_timeout_minutes": 15,
        "pp_size": 1,
        "cp_size": 4,
        "cp_strategy": "cp_ulysses",
        "ulysses_overlap_degree": 1,
        "fp8_quant": true,
        "distill_nearly_clean_chunk_threshold": 0.3,
        "shortcut_mode": "8,16,16",
        "distill": true,
        "kv_offload": false,
        "enable_cuda_graph": true
    }
}
```
```python
def _run(self, prompt: str, prefix_video: torch.Tensor, output_path: str):
        caption_embs, emb_masks = get_txt_embeddings(prompt, self.config)
        dit = get_dit(self.config)
        videos = torch.cat(
            [
                post_chunk_process(chunk, self.config)
                for chunk in generate_per_chunk(
                    model=dit, prompt=prompt, prefix_video=prefix_video, caption_embs=caption_embs, emb_masks=emb_masks
                )
            ],
            dim=0,
        )
        
        self.config.runtime_config.num_frames = 191
        caption_embs, emb_masks = get_txt_embeddings(prompt, self.config)
        dit = get_dit(self.config)
        for iii in range(3):        
            # Start timing video generation if rank 0
            if dist.get_rank() == 0:
                event_path_timer().synced_record("video_generation_start")
                print('Testing inference time cost.')
                
            videos = torch.cat(
                [
                    post_chunk_process(chunk, self.config)
                    for chunk in generate_per_chunk(
                        model=dit, prompt=prompt, prefix_video=prefix_video, caption_embs=caption_embs, emb_masks=emb_masks
                    )
                ],
                dim=0,
            )
            
            # End timing if rank 0
            if dist.get_rank() == 0:
                event_path_timer().synced_record("video_generation_end")
                print('End of testing inference time cost.')
        print(f"videos.shape: {videos.shape}")
        save_video_to_disk(videos, output_path, fps=self.config.runtime_config.fps//3)

        mem_allocated_gb = torch.cuda.max_memory_allocated() / 1024**3
        mem_reserved_gb = torch.cuda.max_memory_reserved() / 1024**3
        print_rank_0(
            f"Finish MagiPipeline, max memory allocated: {mem_allocated_gb:.2f} GB, max memory reserved: {mem_reserved_gb:.2f} GB"
        )

```

**Comment by JohnZhan2023** [comment]
And another two questions are:
1.  why the output video's frame num is changed by the fps in config? I set the frame number to be 47(i2v). If I use 24 as fps, I get 48 frames. If I use 8 as fps, I get 12 frames.
3. why the first time inference costs 7-8s per chunk but later inference only costs 3s per chunk?

**Comment by levi131** [comment]
> Thank you for your wonderful work! I downloaded the distilled and fp8_quant version and use the script below, and I get the result like this (using h200):
> 
>  output_i2v.mp4 
> I think the video result is a little poor and I wonder whether it's correct. I change the frame_num in run_ and my config is:
> 
> {
>     "model_config": {
>         "model_name": "videodit_ardf",
>         "num_layers": 48,
>         "hidden_size": 6144,
>         "ffn_hidden_size": 16384,
>         "num_attention_heads": 48,
>         "num_query_groups": 8,
>         "kv_channels": 128,
>         "layernorm_epsilon": 1e-06,
>         "apply_layernorm_1p": true,
>         "x_rescale_factor": 0.1,
>         "half_channel_vae": true,
>         "params_dtype": "torch.bfloat16",
>         "patch_size": 2,
>         "t_patch_size": 1,
>         "in_channels": 32,
>         "out_channels": 32,
>         "cond_hidden_ratio": 0.25,
>         "caption_channels": 4096,
>         "caption_max_length": 800,
>         "xattn_cond_hidden_ratio": 1.0,
>         "cond_gating_ratio": 1.0,
>         "gated_linear_unit": true
>     },
>     "runtime_config": {
>         "cfg_number": 1,
>         "cfg_t_range": [
>             0.0,
>             0.0217,
>             0.1,
>             0.3,
>             0.999
>         ],
>         "prev_chunk_scales": [
>             1.5,
>             1.5,
>             1.5,
>             1.0,
>             1.0
>         ],
>         "text_scales": [
>             7.5,
>             7.5,
>             7.5,
>             0.0,
>             0.0
>         ],
>         "noise2clean_kvrange": [
>             5,
>             4,
>             3,
>             2
>         ],
>         "clean_chunk_kvrange": 1,
>         "clean_t": 0.9999,
>         "seed": 1234,
>         "num_frames": 23,
>         "video_size_h": 480,
>         "video_size_w": 640,
>         "num_steps": 4,
>         "window_size": 4,
>         "fps": 24,
>         "chunk_width": 6,
>         "load": "./ckpt/magi/24B_distill_quant",
>         "t5_pretrained": "./ckpt/t5",
>         "vae_pretrained": "./ckpt/vae",
>         "scale_factor": 0.18215,
>         "temporal_downsample_factor": 4
>     },
>     "engine_config": {
>         "distributed_backend": "nccl",
>         "distributed_timeout_minutes": 15,
>         "pp_size": 1,
>         "cp_size": 4,
>         "cp_strategy": "cp_ulysses",
>         "ulysses_overlap_degree": 1,
>         "fp8_quant": true,
>         "distill_nearly_clean_chunk_threshold": 0.3,
>         "shortcut_mode": "8,16,16",
>         "distill": true,
>         "kv_offload": false,
>         "enable_cuda_graph": true
>     }
> }
> def _run(self, prompt: str, prefix_video: torch.Tensor, output_path: str):
>         caption_embs, emb_masks = get_txt_embeddings(prompt, self.config)
>         dit = get_dit(self.config)
>         videos = torch.cat(
>             [
>                 post_chunk_process(chunk, self.config)
>                 for chunk in generate_per_chunk(
>                     model=dit, prompt=prompt, prefix_video=prefix_video, caption_embs=caption_embs, emb_masks=emb_masks
>                 )
>             ],
>             dim=0,
>         )
>         
>         self.config.runtime_config.num_frames = 191
>         caption_embs, emb_masks = get_txt_embeddings(prompt, self.config)
>         dit = get_dit(self.config)
>         for iii in range(3):        
>             # Start timing video generation if rank 0
>             if dist.get_rank() == 0:
>                 event_path_timer().synced_record("video_generation_start")
>                 print('Testing inference time cost.')
>                 
>             videos = torch.cat(
>                 [
>                     post_chunk_process(chunk, self.config)
>                     for chunk in generate_per_chunk(
>                         model=dit, prompt=prompt, prefix_video=prefix_video, caption_embs=caption_embs, emb_masks=emb_masks
>                     )
>                 ],
>                 dim=0,
>             )
>             
>             # End timing if rank 0
>             if dist.get_rank() == 0:
>                 event_path_timer().synced_record("video_generation_end")
>                 print('End of testing inference time cost.')
>         print(f"videos.shape: {videos.shape}")
>         save_video_to_disk(videos, output_path, fps=self.config.runtime_config.fps//3)
> 
>         mem_allocated_gb = torch.cuda.max_memory_allocated() / 1024**3
>         mem_reserved_gb = torch.cuda.max_memory_reserved() / 1024**3
>         print_rank_0(
>             f"Finish MagiPipeline, max memory allocated: {mem_allocated_gb:.2f} GB, max memory reserved: {mem_reserved_gb:.2f} GB"
>         )

I think the poor result is caused by the fact that num_steps=4 in your config is too small. We usually use num_steps=16 or 32 in distill mode.

**Comment by levi131** [comment]
> And another two questions are:
> 
> 1. why the output video's frame num is changed by the fps in config? I set the frame number to be 47(i2v). If I use 24 as fps, I get 48 frames. If I use 8 as fps, I get 12 frames.
> 2. why the first time inference costs 7-8s per chunk but later inference only costs 3s per chunk?

For the first question, we often use two setting: (a) fps=24 and chunk_width=6 (b) fps=32 and chunk_width=8. Other setting for example fps=8 and chunk_width=6 may cause incorrect result. For the 24fps and chunk_width=6 setting, because we have some rounding logic in the calculation logic of the final generated video frame number, so the final number of video frames will be slightly different from the input in the config.

**Comment by JohnZhan2023** [comment]
> > Thank you for your wonderful work! I downloaded the distilled and fp8_quant version and use the script below, and I get the result like this (using h200):
> > output_i2v.mp4
> > I think the video result is a little poor and I wonder whether it's correct. I change the frame_num in run_ and my config is:
> > {
> > "model_config": {
> > "model_name": "videodit_ardf",
> > "num_layers": 48,
> > "hidden_size": 6144,
> > "ffn_hidden_size": 16384,
> > "num_attention_heads": 48,
> > "num_query_groups": 8,
> > "kv_channels": 128,
> > "layernorm_epsilon": 1e-06,
> > "apply_layernorm_1p": true,
> > "x_rescale_factor": 0.1,
> > "half_channel_vae": true,
> > "params_dtype": "torch.bfloat16",
> > "patch_size": 2,
> > "t_patch_size": 1,
> > "in_channels": 32,
> > "out_channels": 32,
> > "cond_hidden_ratio": 0.25,
> > "caption_channels": 4096,
> > "caption_max_length": 800,
> > "xattn_cond_hidden_ratio": 1.0,
> > "cond_gating_ratio": 1.0,
> > "gated_linear_unit": true
> > },
> > "runtime_config": {
> > "cfg_number": 1,
> > "cfg_t_range": [
> > 0.0,
> > 0.0217,
> > 0.1,
> > 0.3,
> > 0.999
> > ],
> > "prev_chunk_scales": [
> > 1.5,
> > 1.5,
> > 1.5,
> > 1.0,
> > 1.0
> > ],
> > "text_scales": [
> > 7.5,
> > 7.5,
> > 7.5,
> > 0.0,
> > 0.0
> > ],
> > "noise2clean_kvrange": [
> > 5,
> > 4,
> > 3,
> > 2
> > ],
> > "clean_chunk_kvrange": 1,
> > "clean_t": 0.9999,
> > "seed": 1234,
> > "num_frames": 23,
> > "video_size_h": 480,
> > "video_size_w": 640,
> > "num_steps": 4,
> > "window_size": 4,
> > "fps": 24,
> > "chunk_width": 6,
> > "load": "./ckpt/magi/24B_distill_quant",
> > "t5_pretrained": "./ckpt/t5",
> > "vae_pretrained": "./ckpt/vae",
> > "scale_factor": 0.18215,
> > "temporal_downsample_factor": 4
> > },
> > "engine_config": {
> > "distributed_backend": "nccl",
> > "distributed_timeout_minutes": 15,
> > "pp_size": 1,
> > "cp_size": 4,
> > "cp_strategy": "cp_ulysses",
> > "ulysses_overlap_degree": 1,
> > "fp8_quant": true,
> > "distill_nearly_clean_chunk_threshold": 0.3,
> > "shortcut_mode": "8,16,16",
> > "distill": true,
> > "kv_offload": false,
> > "enable_cuda_graph": true
> > }
> > }
> > def _run(self, prompt: str, prefix_video: torch.Tensor, output_path: str):
> > caption_embs, emb_masks = get_txt_embeddings(prompt, self.config)
> > dit = get_dit(self.config)
> > videos = torch.cat(
> > [
> > post_chunk_process(chunk, self.config)
> > for chunk in generate_per_chunk(
> > model=dit, prompt=prompt, prefix_video=prefix_video, caption_embs=caption_embs, emb_masks=emb_masks
> > )
> > ],
> > dim=0,
> > )
> > ```
> >     self.config.runtime_config.num_frames = 191
> >     caption_embs, emb_masks = get_txt_embeddings(prompt, self.config)
> >     dit = get_dit(self.config)
> >     for iii in range(3):        
> >         # Start timing video generation if rank 0
> >         if dist.get_rank() == 0:
> >             event_path_timer().synced_record("video_generation_start")
> >             print('Testing inference time cost.')
> >             
> >         videos = torch.cat(
> >             [
> >                 post_chunk_process(chunk, self.config)
> >                 for chunk in generate_per_chunk(
> >                     model=dit, prompt=prompt, prefix_video=prefix_video, caption_embs=caption_embs, emb_masks=emb_masks
> >                 )
> >             ],
> >             dim=0,
> >         )
> >         
> >         # End timing if rank 0
> >         if dist.get_rank() == 0:
> >             event_path_timer().synced_record("video_generation_end")
> >             print('End of testing inference time cost.')
> >     print(f"videos.shape: {videos.shape}")
> >     save_video_to_disk(videos, output_path, fps=self.config.runtime_config.fps//3)
> > 
> >     mem_allocated_gb = torch.cuda.max_memory_allocated() / 1024**3
> >     mem_reserved_gb = torch.cuda.max_memory_reserved() / 1024**3
> >     print_rank_0(
> >         f"Finish MagiPipeline, max memory allocated: {mem_allocated_gb:.2f} GB, max memory reserved: {mem_reserved_gb:.2f} GB"
> >     )
> > ```
> 
> I think the poor result is caused by the fact that num_steps=4 in your config is too small. We usually use num_steps=16 or 32 in distill mode.

I changed step to be 16 and get this:

https://github.com/user-attachments/assets/0e7b7744-f92c-45ed-9e52-04e345c74134

Is the result correct? I find some noise and some jittering

**Comment by JohnZhan2023** [comment]
@levi131  also，I wonder whether the distilled model compatible with sdedit? Can I give a reference video using the sdedit method?

**Comment by levi131** [comment]
> caption_embs

I have tried the same config with num_steps=16, num_frames=191, video_size_h=480, video_size_w=640, and could not see noise and jittering, did you make any additional modifications to the code?
My i2v result:

https://github.com/user-attachments/assets/0907c76f-3f3f-4819-991f-cdb76c81b31e
